libraries/AP_Compass/AP_Compass_BMM150.cpp

```diff
@@ -0,0 +1,279 @@
+#define SOFT_RESET (1 << 7 | 1 << 1)
+
+#define OP_MODE_SELF_TEST_ODR_REG 0x4C
+#define NORMAL_MODE (0 << 1)
+#define ODR_30HZ (1 << 3 | 1 << 4 | 1 << 5)
+#define ODR_20HZ (1 << 3 | 0 << 4 | 1 << 5)
+
+#define DATA_X_LSB_REG 0x42
+
+#define REPETITIONS_XY_REG 0x51
+#define REPETITIONS_Z_REG 0X52
+
+/* Trim registers */
+#define DIG_X1_REG 0x5D
+#define DIG_Y1_REG 0x5E
+#define DIG_Z4_LSB_REG 0x62
+#define DIG_Z4_MSB_REG 0x63
+#define DIG_X2_REG 0x64
+#define DIG_Y2_REG 0x65
+#define DIG_Z2_LSB_REG 0x68
+#define DIG_Z2_MSB_REG 0x69
+#define DIG_Z1_LSB_REG 0x6A
+#define DIG_Z1_MSB_REG 0x6B
+#define DIG_XYZ1_LSB_REG 0x6C
+#define DIG_XYZ1_MSB_REG 0x6D
+#define DIG_Z3_LSB_REG 0x6E
+#define DIG_Z3_MSB_REG 0x6F
+#define DIG_XY2_REG 0x70
+#define DIG_XY1_REG 0x71
+
+#define MICRO_SECONDS_BETWEEN_MEASUREMENTS 10000
+
+extern const AP_HAL::HAL &hal;
+
+AP_Compass_Backend *AP_Compass_BMM150::probe(Compass &compass,
+                                             AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+{
+    AP_Compass_BMM150 *sensor = new AP_Compass_BMM150(compass, std::move(dev));
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+AP_Compass_BMM150::AP_Compass_BMM150(Compass &compass,
+                                     AP_HAL::OwnPtr<AP_HAL::Device> dev)
+    : AP_Compass_Backend(compass)
+    , _dev(std::move(dev))
+{
+}
+
+void AP_Compass_BMM150::_load_trim_values()
+{
+    uint8_t data[21];
+
+    _dev->read_registers(DIG_X1_REG, data, sizeof(data));
+
+    _dig_x1 = (int8_t)data[DIG_X1_REG - DIG_X1_REG];
+    _dig_y1 = (int8_t)data[DIG_Y1_REG - DIG_X1_REG];
+    _dig_x2 = (int8_t)data[DIG_X2_REG - DIG_X1_REG];
+    _dig_y2 = (int8_t)data[DIG_Y2_REG - DIG_X1_REG];
+    _dig_xy1 = data[DIG_XY1_REG - DIG_X1_REG];
+    _dig_xy2 = (int8_t)data[DIG_XY2_REG - DIG_X1_REG];
+    _dig_z1 = (uint16_t)le16toh(data[DIG_Z1_LSB_REG - DIG_X1_REG] | data[DIG_Z1_MSB_REG - DIG_X1_REG] << 8);
+    _dig_z2 = (int16_t)le16toh(data[DIG_Z2_LSB_REG - DIG_X1_REG] | data[DIG_Z2_MSB_REG - DIG_X1_REG] << 8);
+    _dig_z3 = (int16_t)le16toh(data[DIG_Z3_LSB_REG - DIG_X1_REG] | data[DIG_Z3_MSB_REG - DIG_X1_REG] << 8);
+    _dig_z4 = (int16_t)le16toh(data[DIG_Z4_LSB_REG - DIG_X1_REG] | data[DIG_Z4_MSB_REG - DIG_X1_REG] << 8);
+    _dig_xyz1 = (uint16_t)le16toh(data[DIG_XYZ1_LSB_REG - DIG_X1_REG] | data[DIG_XYZ1_MSB_REG - DIG_X1_REG] << 8);
+}
+
+bool AP_Compass_BMM150::init()
+{
+    hal.scheduler->suspend_timer_procs();
+
+    if (!_dev->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        hal.console->printf("BMM150: Unable to get bus semaphore\n");
+        goto fail_sem;
+    }
+
+    /* Do a soft reset */
+    _dev->write_register(POWER_AND_OPERATIONS_REG, SOFT_RESET);
+    hal.scheduler->delay(2);
+
+    /* Change power state from suspend mode to sleep mode */
+    _dev->write_register(POWER_AND_OPERATIONS_REG, POWER_CONTROL_VAL);
+    hal.scheduler->delay(2);
+
+    uint8_t val;
```

Comment:
Initialize Val to 0, so if the read_registers() call fails you don't compare the return with garbage